rs/sns/governance/src/proposal.rs

```diff
@@ -2770,78 +2774,268 @@ mod tests {
+            .unwrap();
 
         assert_eq!(
             text,
-            r#"# Proposal to upgrade SNS controlled canister:
+            r#"# Proposal to Upgrade an SNS Controlled Canister
 
-## Canister id: bg4sm-wzk
+## Target canister: xbgkv-fyaaa-aaaaa-aaava-cai
 
-## Canister wasm sha256: 93a44bbb96c751218e4c00d479e4c14358122a389acca16205b1e4d0dc5f9476
+## Wasm info
+
+Embedded module with 8 bytes and SHA256 `93a44bbb96c751218e4c00d479e4c14358122a389acca16205b1e4d0dc5f9476`.
 
 ## Mode: Upgrade
 
-## No upgrade arg"#
+## Argument info
+
+No upgrade argument."#
                 .to_string()
         );
     }
 
-    #[test]
-    fn render_upgrade_sns_controlled_canister_proposal_with_upgrade_args() {
+    #[tokio::test]
+    async fn render_upgrade_sns_controlled_canister_proposal_with_chunked_wasm() {
+        let upgrade = UpgradeSnsControlledCanister {
+            canister_id: Some(basic_canister_id()),
+            new_canister_wasm: vec![],
+            canister_upgrade_arg: None,
+            mode: Some(CanisterInstallModeProto::Upgrade.into()),
+            chunked_canister_wasm: Some(ChunkedCanisterWasm {
+                wasm_module_hash: vec![1, 2, 3],
+                store_canister_id: Some(canister_test_id(111).get()),
+                chunk_hashes_list: vec![vec![1, 1, 1], vec![2, 2, 2], vec![3, 3, 3]],
+            }),
+        };
+        let env = setup_for_upgrade_sns_controlled_canister_tests(&upgrade);
+        let text = validate_and_render_upgrade_sns_controlled_canister(&upgrade, &env)
+            .await
+            .unwrap();
+
+        assert_eq!(
+            text,
+            r#"# Proposal to Upgrade an SNS Controlled Canister
+
+## Target canister: xbgkv-fyaaa-aaaaa-aaava-cai
+
+## Wasm info
+
+Remote module stored on canister zyo6l-paaaa-aaaaa-aabxq-cai with SHA256 `010203`. Split into 3 chunks:
+  - `010101`
+  - `020202`
+  - `030303`
+
+## Mode: Upgrade
+
+## Argument info
+
+No upgrade argument."#
+                .to_string()
+        );
+    }
+
+    #[tokio::test]
+    async fn render_upgrade_sns_controlled_canister_proposal_with_unexpected_chunk() {
+        let mut chunked_canister_wasm = ChunkedCanisterWasm {
+            wasm_module_hash: vec![1, 2, 3],
+            store_canister_id: Some(canister_test_id(111).get()),
+            chunk_hashes_list: vec![vec![1, 1, 1], vec![2, 2, 2], vec![3, 3, 3]],
+        };
+        let mut upgrade = UpgradeSnsControlledCanister {
+            canister_id: Some(basic_canister_id()),
+            new_canister_wasm: vec![],
+            canister_upgrade_arg: None,
+            mode: Some(CanisterInstallModeProto::Upgrade.into()),
+            chunked_canister_wasm: Some(chunked_canister_wasm.clone()),
+        };
+
+        let env = setup_for_upgrade_sns_controlled_canister_tests(&upgrade);
+
+        // Modify the update payload to make it invalid (unexpected chunk).
+        {
+            chunked_canister_wasm.chunk_hashes_list.push(vec![4, 4, 4]);
```

Comment:
No change required.

Why not simply do this on line 2855?

Ditto elsewhere.

Reply:
Because we make an `env` out of the first version of `upgrade`, and then change `upgrade`.